### PR TITLE
My Site Dashboard-Phase 2: Implement pull-to-refresh

### PR DESCRIPTION
**Pull request by zwarm**
Parent #15215 

This is the final PR in a series (see parent issue)  that refactored, prepared, and implemented end-to-end pull to refresh for the My Site tab. The PTR logic is controlled by the MySiteDashboardPhase2FeatureConfig feature flag, thus minimizing risk to existing functionality.

This PR focuses on the following:
- Ensure that `refresh()` is only called for sources that implement `MySiteRefreshSource`
- Hide the refresh indicator when all sources have finished refreshing
- Kick off `refresh()` when a pull-to-refresh occurs
- Kick off `refresh()` during `onResume` event except for initial request

#### **Test 1**
- Launch the app
- Navigate to My Site -> App Settings -> Debug Settings
- Ensure that the `MySiteDashboardPhase2FeatureConfig` is to set to off
- Restart the app if needed
- Navigate to the My Site tab
- Try to swiping down on the view to trigger the refresh
- Note: swipe to refresh is not active

#### **Test 2**
- Launch the app
- Navigate to My Site -> App Settings -> Debug Settings
- Set the `MySiteDashboardPhase2FeatureConfig` flag to on
- Restart the App
- Navigate to the My Site tab
- Try to swiping down on the view to trigger the refresh
- Note: 
      The swipe to refresh indicator is shown at first
      The "upcoming scheduled posts" posts cards data will change to say "refreshed" ( they are hardcoded to change)
      Thee  swipe to refresh indicator is no longer shown when all data has reloaded

## Regression Notes
1. Potential unintended areas of impact
The cards don't refresh when requested

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual testing and new automated tests

3. What automated tests I added (or what prevented me from doing so)
New tests in MySiteViewModelTest

PR submission checklist:
- [X] I have completed the Regression Notes.
- [X] I have considered adding accessibility improvements for my changes.
- [X] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WordPress/src/main/java/org/wordpress/android/ui/mysite/MySiteViewModel.kt [hunk 1]
```diff
@@ -738,15 +745,29 @@ class MySiteViewModel @Inject constructor(
     }
 
     fun onPullToRefresh() {
-        postCardsSource.refresh()
+        refresh()
```
Comment (ParaskP7) [inline WordPress/src/main/java/org/wordpress/android/ui/mysite/MySiteViewModel.kt]: Suggestion (💡): Consider removing this extra `onPullToRefresh()` function from this view model and use directly the `refresh()`. This way, you wouldn't at least need to create another set of test for this publicly available function. Or is it that this is needed because in the future it will be doing more than just calling `refresh()`?

File: WordPress/src/main/java/org/wordpress/android/ui/mysite/MySiteFragment.kt [hunk 2]
```diff
@@ -519,10 +526,11 @@ class MySiteFragment : Fragment(R.layout.my_site_fragment),
 
     private fun MySiteFragmentBinding.loadData(cardAndItems: List<MySiteCardAndItem>) {
         recyclerView.setVisible(true)
+        actionableEmptyView.setVisible(false)
+        hideRefreshIndicatorIfNeeded()
```
Comment (ParaskP7) [inline WordPress/src/main/java/org/wordpress/android/ui/mysite/MySiteFragment.kt]: Suggestion (🔍): Consider moving this up a level. Maybe this doesn't belong to the `loadData(...)` method because it is not loading any data, wdyt?
```
            when (val state = uiModel.state) {
                is State.SiteSelected -> {
                    hideRefreshIndicatorIfNeeded()
                    loadData(state.cardAndItems)
                }
                is State.NoSites -> loadEmptyView(state.shouldShowImage)
            }
```

File: WordPress/src/main/java/org/wordpress/android/ui/mysite/MySiteViewModel.kt [hunk 3]
```diff
@@ -482,9 +484,13 @@ class MySiteViewModel @Inject constructor(
     }
 
     fun refresh() {
-        selectedSiteRepository.updateSiteSettingsIfNecessary()
-        quickStartCardSource.refresh()
-        currentAvatarSource.refresh()
+        if (mySiteDashboardPhase2FeatureConfig.isEnabled()) {
+            mySiteSources.filterIsInstance(MySiteRefreshSource::class.java).forEach { it.refresh() }
+        } else {
+            selectedSiteRepository.updateSiteSettingsIfNecessary()
```
Comment (ParaskP7) [inline WordPress/src/main/java/org/wordpress/android/ui/mysite/MySiteViewModel.kt]: Question (❓): Shouldn't this `selectedSiteRepository.updateSiteSettingsIfNecessary()` be called even if `mySiteDashboardPhase2FeatureConfig` is enabled? For example something like the below:

```
    fun refresh() {
        selectedSiteRepository.updateSiteSettingsIfNecessary()
        if (mySiteDashboardPhase2FeatureConfig.isEnabled()) {
            mySiteSources.filterIsInstance(MySiteRefreshSource::class.java).forEach { it.refresh() }
        } else {
            quickStartCardSource.refresh()
            currentAvatarSource.refresh()
        }
    }
```

File: WordPress/src/main/java/org/wordpress/android/ui/mysite/MySiteViewModel.kt [hunk 4]
```diff
@@ -670,6 +676,7 @@ class MySiteViewModel @Inject constructor(
         domainRegistrationSource.clear()
         quickStartRepository.clear()
         scanAndBackupSource.clear()
+        selectedSiteSource.clear()
```
Comment (ParaskP7) [inline WordPress/src/main/java/org/wordpress/android/ui/mysite/MySiteViewModel.kt:701]: Suggestion (💡): Maybe it is time we add tests for this `onCleared()` function, wdyt?

File: WordPress/src/main/java/org/wordpress/android/ui/mysite/MySiteViewModel.kt [hunk 5]
```diff
@@ -738,15 +745,29 @@ class MySiteViewModel @Inject constructor(
     }
 
     fun onPullToRefresh() {
-        postCardsSource.refresh()
+        refresh()
     }
 
+    @Suppress("unused", "UNUSED_PARAMETER")
```
Comment (ParaskP7) [inline WordPress/src/main/java/org/wordpress/android/ui/mysite/MySiteViewModel.kt]: Minor (🔍): Consider removing the `unused` as it is actually used. However, the parameter it is indeed unused and as such the `UNUSED_PARAMETER` can stay.

File: WordPress/src/test/java/org/wordpress/android/ui/mysite/MySiteViewModelTest.kt [hunk 6]
```diff
@@ -443,6 +463,42 @@ class MySiteViewModelTest : BaseUnitTest() {
...
+        allRefreshedMySiteSources.filterIsInstance(MySiteRefreshSource::class.java).forEach { verify(it).refresh() }
+    }
+
+    @Test
+    fun `given mySiteDashboardPhase2FeatureConfig is disabled, when refresh triggers, then select sources refresh`() {
+        whenever(mySiteDashboardPhase2FeatureConfig.isEnabled()).thenReturn(false)
+
+        viewModel.refresh()
+
+        selectRefreshedMySiteSources.filterIsInstance(MySiteRefreshSource::class.java).forEach { verify(it).refresh() }
```
Comment (ParaskP7) [inline WordPress/src/test/java/org/wordpress/android/ui/mysite/MySiteViewModelTest.kt:532]: Minor (🔍): Consider also checking for `        verify(selectedSiteRepository).updateSiteSettingsIfNecessary()`, either within this test or a separate test that tests that in isolation.

PS: As per my previous comment, maybe this `selectedSiteRepository.updateSiteSettingsIfNecessary()` should be called even when `mySiteDashboardPhase2FeatureConfig` is enabled. 🤔

File: WordPress/src/test/java/org/wordpress/android/ui/mysite/MySiteViewModelTest.kt [hunk 7]
```diff
@@ -443,6 +463,42 @@ class MySiteViewModelTest : BaseUnitTest() {
...
+    }
+
+    @Test
+    fun `given mySiteDashboardPhase2FeatureConfig is enabled, when refreshing, then refresh indicator should show`() {
```
Comment (ParaskP7) [inline WordPress/src/test/java/org/wordpress/android/ui/mysite/MySiteViewModelTest.kt]: 1. Minor (🔍): Consider changing the `given ... , when refreshing, then ...` to `given ... , when are sources refreshing, then ...`
2. Suggestion (💡): Consider rewriting this test with a `given/when/then` structure, like the below:

```
    @Test
    fun `given mySiteDashboardPhase2FeatureConfig is enabled, when are sources refreshing, then refresh indicator should show`() {
        whenever(mySiteDashboardPhase2FeatureConfig.isEnabled()).thenReturn(true)
        allRefreshedMySiteSources.filterIsInstance(MySiteRefreshSource::class.java).forEach {
            whenever(it.isRefreshing()).thenReturn(true)
        }

        val result = viewModel.shouldHideRefreshIndicator()

        assertFalse(result)
    }
```

3. Suggestion (💡): Consider writing another accompanying test for the case when `whenever(it.isRefreshing()).thenReturn(false)` to cover both cases.

File: WordPress/src/test/java/org/wordpress/android/ui/mysite/MySiteViewModelTest.kt [hunk 8]
```diff
@@ -443,6 +463,42 @@ class MySiteViewModelTest : BaseUnitTest() {
...
+
+        allRefreshedMySiteSources.filterIsInstance(MySiteRefreshSource::class.java).forEach {
+            whenever(it.isRefreshing()).thenReturn(true)
+        }
+
+        assertThat(viewModel.shouldHideRefreshIndicator()).isFalse
+    }
+
+    @Test
+    fun `when mySiteDashboardPhase2FeatureConfig is disabled, then refresh indicator should hide`() {
```
Comment (ParaskP7) [inline WordPress/src/test/java/org/wordpress/android/ui/mysite/MySiteViewModelTest.kt]: Suggestion (💡): Consider rewriting this test with a given/when/then structure, like the below:

```
    @Test
    fun `given mySiteDashboardPhase2FeatureConfig is disabled, when are sources refreshing, then refresh indicator should hide`() {
        whenever(mySiteDashboardPhase2FeatureConfig.isEnabled()).thenReturn(false)

        val result = viewModel.shouldHideRefreshIndicator()

        assertTrue(result)
    }
```

File: WordPress/src/main/java/org/wordpress/android/ui/mysite/MySiteViewModel.kt [hunk 9]
```diff
@@ -738,15 +745,29 @@ class MySiteViewModel @Inject constructor(
...
     fun onPostCardFooterLinkClick(postCardType: PostCardType) {
         selectedSiteRepository.getSelectedSite()?.let {
             // TODO: ashiagr implement navigation
         }
     }
 
+    fun shouldHideRefreshIndicator() = haveAllSourcesFinishedRefreshing()
+
+    private fun haveAllSourcesFinishedRefreshing(): Boolean {
```
Comment (ParaskP7) [inline WordPress/src/main/java/org/wordpress/android/ui/mysite/MySiteViewModel.kt]: Suggestion (💡): What if this method was renaming to a simpler `areSourcesRefreshing()`, make public, and then this is used directly instead of having the `shouldHideRefreshIndicator()` intermediate?

**Review by ParaskP7 — CHANGES_REQUESTED** [review]
👋 @zwarm !

I have reviewed and tested this PR as per the instructions, everything kind-of works as expected, good job! 🌟 

-----

I have left a question (❓), some suggestions (💡) and a few minor (🔍) comments for you to consider. I am **NOT** going to approve this PR for now, since I have experienced few bugs that I want to discuss with you first (see `EXTRAS` below).

-----

`EXTRAS`
1. Warning (⚠️): During testing I would reproduce the `PTR` to refresh indefinitely. Did you also get to experience that during testing? Let's discuss.
2. Warning (⚠️): During testing I found an inconsistency with `Quick Start` not showing (when enabled) when the feature flag is off, while it is showing when the feature flag is on with `PTR`. Did you notice that?
3. Suggestion (💡): I was thinking about extracting all the `Source` related functionality into a `Manager` class, maybe calling it a `MySiteSourceManager`. I will share a patch with you to discuss. PS: If you like it, this refactor can be done in a separate PR.

File: WordPress/src/main/java/org/wordpress/android/ui/mysite/MySiteViewModel.kt (hunk 3 above)
Comment (zwarm) [reply WordPress/src/main/java/org/wordpress/android/ui/mysite/MySiteViewModel.kt]: Nope, because it's part of refresh in `SelectedSiteSource` . This stays as is.

File: WordPress/src/main/java/org/wordpress/android/ui/mysite/MySiteViewModel.kt (hunk 1 above)
Comment (zwarm) [reply WordPress/src/main/java/org/wordpress/android/ui/mysite/MySiteViewModel.kt]: Let's change it to refresh, but reserve the right to create a new fun if we need to differentiate between user requested PTR and automatic.

File: WordPress/src/main/java/org/wordpress/android/ui/mysite/MySiteFragment.kt (hunk 2 above)
Comment (zwarm) [reply WordPress/src/main/java/org/wordpress/android/ui/mysite/MySiteFragment.kt]: Moving it out into setupObservers

File: WordPress/src/main/java/org/wordpress/android/ui/mysite/MySiteViewModel.kt (hunk 4 above)
Comment (zwarm) [reply WordPress/src/main/java/org/wordpress/android/ui/mysite/MySiteViewModel.kt:701]: Absolutely can add them :)

File: WordPress/src/main/java/org/wordpress/android/ui/mysite/MySiteViewModel.kt (hunk 5 above)
Comment (zwarm) [reply WordPress/src/main/java/org/wordpress/android/ui/mysite/MySiteViewModel.kt:751]: will do

File: WordPress/src/test/java/org/wordpress/android/ui/mysite/MySiteViewModelTest.kt (hunk 7 above)
Comment (zwarm) [reply WordPress/src/test/java/org/wordpress/android/ui/mysite/MySiteViewModelTest.kt]: Just me saving lines, I'll break into two. :(

File: WordPress/src/test/java/org/wordpress/android/ui/mysite/MySiteViewModelTest.kt (hunk 8 above)
Comment (zwarm) [reply WordPress/src/test/java/org/wordpress/android/ui/mysite/MySiteViewModelTest.kt:496]: Just me saving lines, I'll break into two. :(

File: WordPress/src/main/java/org/wordpress/android/ui/mysite/MySiteViewModel.kt (hunk 9 above)
Comment (zwarm) [reply WordPress/src/main/java/org/wordpress/android/ui/mysite/MySiteViewModel.kt]: Sure. :)

File: WordPress/src/test/java/org/wordpress/android/ui/mysite/MySiteViewModelTest.kt (hunk 6 above)
Comment (zwarm) [reply WordPress/src/test/java/org/wordpress/android/ui/mysite/MySiteViewModelTest.kt:532]: See previous comment about this functionality being moved to selectedSiteSource post flag turned on. Test added in b57a1e4

File: WordPress/src/test/java/org/wordpress/android/ui/mysite/MySiteViewModelTest.kt (hunk 8 above)
Comment (zwarm) [reply WordPress/src/test/java/org/wordpress/android/ui/mysite/MySiteViewModelTest.kt]: The test `when mySiteDashboardPhase2FeatureConfig is disabled, then refresh indicator should hide` is for when the  flag is disabled - meaning sources can't be refreshed, because PTR isn't enabled. (unless of course a source is individually refreshed - interesting, this may be the refresh issue you stumbled on - completely unrelated to this test here)

File: WordPress/src/test/java/org/wordpress/android/ui/mysite/MySiteViewModelTest.kt (hunk 7 above)
Comment (zwarm) [reply WordPress/src/test/java/org/wordpress/android/ui/mysite/MySiteViewModelTest.kt]: in 340192b - although the fun name is too long to pass detekt

File: WordPress/src/main/java/org/wordpress/android/ui/mysite/MySiteViewModel.kt (hunk 1 above)
Comment (zwarm) [reply WordPress/src/main/java/org/wordpress/android/ui/mysite/MySiteViewModel.kt]: in c67b923

File: WordPress/src/main/java/org/wordpress/android/ui/mysite/MySiteFragment.kt (hunk 2 above)
Comment (zwarm) [reply WordPress/src/main/java/org/wordpress/android/ui/mysite/MySiteFragment.kt]: in a147c0e

File: WordPress/src/main/java/org/wordpress/android/ui/mysite/MySiteViewModel.kt (hunk 4 above)
Comment (zwarm) [reply WordPress/src/main/java/org/wordpress/android/ui/mysite/MySiteViewModel.kt:701]: in 176d9a0

File: WordPress/src/main/java/org/wordpress/android/ui/mysite/MySiteViewModel.kt (hunk 5 above)
Comment (zwarm) [reply WordPress/src/main/java/org/wordpress/android/ui/mysite/MySiteViewModel.kt]: Removed in febc6f9

File: WordPress/src/main/java/org/wordpress/android/ui/mysite/MySiteViewModel.kt (hunk 9 above)
Comment (zwarm) [reply WordPress/src/main/java/org/wordpress/android/ui/mysite/MySiteViewModel.kt]: in c67b923

File: WordPress/src/main/java/org/wordpress/android/ui/mysite/MySiteViewModel.kt (hunk 3 above)
Comment (ParaskP7) [reply WordPress/src/main/java/org/wordpress/android/ui/mysite/MySiteViewModel.kt]: True, many thanks for the explanation Annmarie! 🙏

**Comment by zwarm** [comment]
@ParaskP7 - thanks for the great review! See below for some more info. :)

> Warning (⚠️): During testing I would reproduce the PTR to refresh indefinitely. Did you also get to experience that during testing? Let's discuss. 

Steps to reproduce
- Launch app
- PTR
- Refresh indicator shows & hides
- Swipe back to leave app
- Relaunch app
- PTR
- Refresh indicator shows indefinitely

I have to do so more research on this one, but the gist of it is after we dispatch the call, we don't get an event back, which means the source stays in refresh mode indefinitely. I wonder if there is a "time" delay for activating the same request multiple times? Going to need more time on this one for sure, but it shouldn't be a blocker for this PR since it's behind the feature flag. I haven't seen this happen on launch or relaunch when the flag is disabled, that was your experience as well, correct?.

**UPDATE**:  Fix in baaefa1.  Had to remove `singleton` annotation because eventbus was being unregistered. Cool find!!. Chat  tomorrow if you want extra details. 👍 

>  Warning (⚠️): During testing I found an inconsistency with Quick Start not showing (when enabled) when the feature flag is off, while it is showing when the feature flag is on with PTR. Did you notice that?

This was caused by onResume not executing "non-sources/sources" on a first time onResume.  This has been refactored in edf7b18. `MySiteFragment.onResume()` now passes control over to `MySiteViewModel`. It's the VM job to properly act upon first call or subsequent calls. The `mySiteDashboardPhase2FeatureConfig` has moved from `refresh()` to `onResume()`. The `refresh()` method is only invoked from the PTR (which is locked when the flag is disabled). Much easier to test now that `onResume` is in the VM. I added a bunch of tests.

>  Suggestion (💡): I was thinking about extracting all the Source related functionality into a Manager class, maybe calling it a MySiteSourceManager. I will share a patch with you to discuss. PS: If you like it, this refactor can be done in a separate PR.

As discussed, this will be incorporated as a separate PR.

File: WordPress/src/test/java/org/wordpress/android/ui/mysite/MySiteViewModelTest.kt (hunk 8 above)
Comment (zwarm) [reply WordPress/src/test/java/org/wordpress/android/ui/mysite/MySiteViewModelTest.kt]: Nope, not the issue you ran into - that is all on SelectedSiteSource and that's a whole other discussion. lol.This test stands as written.

File: WordPress/src/test/java/org/wordpress/android/ui/mysite/MySiteViewModelTest.kt (hunk 7 above)
Comment (zwarm) [reply WordPress/src/test/java/org/wordpress/android/ui/mysite/MySiteViewModelTest.kt]: 🤔  It's late and I'm drawing a blank on the correct way to break this test name onto multiple lines. Tomorrow.

**Comment by ParaskP7** [comment]
👋 @zwarm !

> UPDATE: Fix in baaefa1. Had to remove singleton annotation because eventbus was being unregistered. Cool find!!. Chat tomorrow if you want extra details. 👍

> This was caused by onResume not executing "non-sources/sources" on a first time onResume. This has been refactored in edf7b18. 

Thanks for all the fixes you did, I reviewed all that once more and can verify that those problems are now gone. However, I noted the below:
1. Warning (⚠️): The `PTR` can be now activated even when the `Phase 2` flag is disabled. You just have to open the app, go to another tab, come back, and try to pull-to-refresh. You will notice that the `PTR` get available. Can you confirm that?
2. Warning (⚠️): You refactored the `refresh()` functionality and removed the `if (mySiteDashboardPhase2FeatureConfig.isEnabled())` logic. However, in addition to the `viewModel.refresh()` during `swipeToRefreshHelper` on the `MySiteFragment` side, this `refresh()` function is also called from within the `MySiteViewModel.onRemoveNextStepsDialogPositiveButtonClicked()` as well, which means for for when the `Phase 2` is disabled this will misbehave, right?
3. Question (❓): I noticed the removal of `@Singleton` from the `SelectedSiteSource` to fix the issue, as you mentioned. And now I am thinking, should we remove that for all sources? I am not sure anymore why some of our sources are singletons, while others are not. I suggest we remove the `@Singleton` from all our sources and make this consistent everywhere and more deterministic this way. Wdyt?
4. Praise (❤️): I like what you did with moving the `onResume()` logic to view model and then extensively tested it, kudos.
5. Suggestion (💡): On the above note, I created and will be sharing a patch with you on making this view model `onResume()` logic less complex and fixing the tests as well. Let's discuss on that patch so that I explain to you all the improvements I did, including, but not limited to fixing the warning no.2 above on this patch.

**Comment by zwarm** [comment]
> Warning (⚠️): The PTR can be now activated even when the Phase 2 flag is disabled. You just have to open the app, go to another tab, come back, and try to pull-to-refresh. You will notice that the PTR get available. Can you confirm that?

🤔 , that shouldn't be happening because the liveData is triggered directly off the flag itself. I'll see if that's one of those cases where you really need to close the app when flipping the flag.

> Warning (⚠️): You refactored the refresh() functionality and removed the if (mySiteDashboardPhase2FeatureConfig.isEnabled()) logic. However, in addition to the viewModel.refresh() during swipeToRefreshHelper on the MySiteFragment side, this refresh() function is also called from within the MySiteViewModel.onRemoveNextStepsDialogPositiveButtonClicked() as well, which means for for when the Phase 2 is disabled this will misbehave, right?

You are correct here, very good catch. I'll adjust the refresh logic in the VM.

> Question (❓): I noticed the removal of @Singleton from the SelectedSiteSource to fix the issue, as you mentioned. And now I am thinking, should we remove that for all sources? I am not sure anymore why some of our sources are singletons, while others are not. I suggest we remove the @Singleton from all our sources and make this consistent everywhere and more deterministic this way. Wdyt?

The singleton causes issues with the event bus, so any source that relies on it should not be a singleton. I thought I already checked other "dispatcher" sources for the annotation, but I will  double check. I don't believe there is an issue leaving singleton on predefined classes. `SelectedSiteSource` was a break down from `MySiteSource` which was a singleton. 

> Suggestion (💡): On the above note, I created and will be sharing a patch with you on making this view model onResume() logic less complex and fixing the tests as well. Let's discuss on that patch so that I explain to you all the improvements I did, including, but not limited to fixing the warning no.2 above on this patch.

Thanks for the patch, as I appreciate seeing your pov.

**Comment by zwarm** [comment]
Patch applied in 9df2915 @ParaskP7

File: WordPress/src/test/java/org/wordpress/android/ui/mysite/MySiteViewModelTest.kt (hunk 7 above)
Comment (zwarm) [reply WordPress/src/test/java/org/wordpress/android/ui/mysite/MySiteViewModelTest.kt]: @ParaskP7 - ha, you changed the text of these tests in that patch and shortened the name. All good as of 9df2915

**Comment by zwarm** [comment]
> Warning (⚠️): The PTR can be now activated even when the Phase 2 flag is disabled. You just have to open the app, go to another tab, come back, and try to pull-to-refresh. You will notice that the PTR get available. Can you confirm that?

I moved the posting of the value over to `onResume` to prohibit this. In a725f95 with tests in 289e998. Working much better  now. :) This was a great find @ParaskP7 🙇 

I could also have added a method to the VM for isEnabled and call that directly. Either solution would be okay and am willing to change.

File: WordPress/src/test/java/org/wordpress/android/ui/mysite/MySiteViewModelTest.kt (hunk 7 above)
Comment (ParaskP7) [reply WordPress/src/test/java/org/wordpress/android/ui/mysite/MySiteViewModelTest.kt]: Thank you @zwarm !

> ha, you changed the text of these tests in that patch and shortened the name.

Yes, you don't need to have the full feature flag name, just a name that you can distinguish is enough. This is what made the difference... 😅

**Comment by ParaskP7** [comment]
👋 @zwarm !

I tested the solution after applying your latest commits and everything seems to be working okay at the moment, I couldn't break it this time, good job! 🌟

Btw, from me it is a 🟢 light, but maybe this is a good opportunity for Meaghan to thoroughly test it as well before merging to `develop`, especially more towards the flow when the `Phase 2` flag is disabled. This is what our users are going to experience in the next release, which is going to last a bit longer due to the `Christmas/New Year` period. Wdyt?

> The singleton causes issues with the event bus, so any source that relies on it should not be a singleton. I thought I already checked other "dispatcher" sources for the annotation, but I will double check. I don't believe there is an issue leaving singleton on predefined classes. SelectedSiteSource was a break down from MySiteSource which was a singleton.

About this, wdyt about doing another PR afterwards, which will remove the `@Singleton` from all sources so we can test it and see if everything works as expected? I am not in favor of having both singleton and non-singleton sources, especially without us knowing why is that. I very much prefer if that was one way or another, there is no point in mixing and matching just because it works, it feels non-deterministic and we might end up having problems going forward with that. New sources will be added, either as singletons or not, just because of a copy-paste from a sibling source. Your thoughts? PS: Mind the fact that going forward more and more team will end up contributing to the feed, adding feed capabilities based on their features, and this can very quickly become very messy if we don't lower the complexity and make things more deterministic development-wise.

**Review by ParaskP7 — APPROVED** [review]

**Comment by zwarm** [comment]
@ParaskP7 🙇‍♀️  Super appreciate all the conversations we've had this week. 

> About this, wdyt about doing another PR afterwards, which will remove the @Singleton from all sources so we can test it and see if everything works as expected?

We are on the same wave length, this is on my task list for right after MySiteSourceManager!!!

+1 on asking @startuptester - if you have time, do you want to take a look at this one. Basically seeing if all is working as expected. Please and thank you. I'll add you as a reviewer.

**Comment by ParaskP7** [comment]
👋 @zwarm !

> @ParaskP7 🙇‍♀️ Super appreciate all the conversations we've had this week.

Same! ❤️ 

> We are on the same wave length, this is on my task list for right after MySiteSourceManager!!!

Perfecto! 🥇 

> +1 on asking @startuptester - if you have time, do you want to take a look at this one. Basically seeing if all is working as expected. Please and thank you. I'll add you as a reviewer.

Awesome! 🌟

**Comment by startuptester** [comment]
Tested the apk on a Samsung S20 5G
mostly LGTM, a few things to bring up though: 
1. Navigating between  My Site/ Reader or My Site/ Notifs, the My Site screen visibly refreshes automatically (dims for a second) without the refresh spinner/ indicator. I compared this with the 18.5 prod version and do not see it happening there. Is this intentional? 

2. PTR > Switch Sites > PTR - refresh hangs. This happened with every site I switched to. Did anyone else experience this?

**Comment by zwarm** [comment]
Thanks @startuptester. 👍 

Can I ask for a video of both instances, as I am not able to reproduce either of these issues on my S20 or Pixel. Thanks 🙏 

> 1. Navigating between  My Site/ Reader or My Site/ Notifs, the My Site screen visibly refreshes automatically (dims for a second) without the refresh spinner/ indicator. I compared this with the 18.5 prod version and do not see it happening there. Is this intentional?

When the feature flag is on, then yes, the screen will refresh without a spinner/indicator on a return to the my site tab. This is intentional because we need to capture the possibility of changed posts. ❓ Do you encounter this behavior when the feature flag is off?

> 2. PTR > Switch Sites > PTR - refresh hangs. This happened with every site I switched to. Did anyone else experience this? 

Nope, I fired up my S20 too and was not able to reproduce. Again, these features are behind a flag. That's why I asked for a video. Thanks!!

**Comment by startuptester** [comment]
Thanks @zwarm 😁 

> When the feature flag is on, then yes, the screen will refresh without a spinner/indicator on a return to the my site tab. 

All good here then. 

 I captured the hang and actually a crash while site switching during the refresh. Am I doing something wrong? Feature flag is on, otherwise there would be no PTR :) 
Network is good. 

https://user-images.githubusercontent.com/66104034/143974304-ad41fe07-15e4-49bb-948f-3249719a2492.mp4
 
Here's where I am after further investigation. I did this on a fresh install. 
1) login and select a site, choose no quick start help
2) On My Site page, PTR
3) PTR again, spins for at least 5 minutes - I gave up at that point so I'm not sure it ever ends
4) Publish a new post
5) go back to My Site and PTR 
If you do this before toast clears, it spins forever 
If you PTR after the toast is gone, it refreshes. 
6) Subsequent PTRs spin and spin if there is nothing to update. 

Are you able to repro that? I did it several times. I can record that whole thing if you need, but I need to charge this first. 😄

**Comment by zwarm** [comment]
@startuptester 

I have not been able to reproduce your issue at all. This PR did have some troubles with forever PTR early on, but those were all addressed and confirmed as working. 

What I did this morning was: uninstall all WP APK's on my S20, download the [APK](https://github.com/wordpress-mobile/WordPress-Android/pull/15601#issuecomment-974305148) from this PR to my S20, fresh install, and followed your steps exactly. No luck in recreating this issue.

@ParaskP7 were you able to recreate this?
@startuptester - thanks so much for your patience and assistance 🙇  Can you attach the crash error log? That would be helpful. Even better, can you grab the APK from this pr #15620 and try that one? That PR builds off of this one. I wonder if this is just a matter of an old APK (really grasping here). Thanks!!!

**Comment by zwarm** [comment]
@startuptester & @ParaskP7 -  after many, many, many controlled random testing scenarios I have ***finally*** found something. What triggered it was switching sites to-from non-free sites - or a.k.a. ones that trigger a fetch to determine if domain registration card should be shown and then doing a PTR. 

There is an issue with `DomainRegistrationSource` and the way it is fetching the data. The source uses continuation, but you can only resume a continuation one time and we are trying to resume continuation more than once. We end up with (1) a continuation cancel exception and (2) a illegalStateException (I believe this is the crash that @startuptester experienced) This type of logic worked perfectly fine for non-PTR situations, but once we added the `refreshData` support, the source falls on its face when switching sites.  

In addition to the above issue, there is also some weirdness with `distinctUntilChanged()`. I'm still not 100% convinced we need to remove this and I'd like to address the issues with the DomainRegistrationSource first. 

TLDR; Since the next PR #15620 builds off of this PR, but also changes how the sources are managed. I'd like to suggest that we merge this PR and the next. Following that I can open a brand new issue to address the `DomainRegistrationSource` changes. It's becoming too unwieldy to keep this PR up-to-date with develop + everything PTR related is behind the feature flag.

So .. I need the 👍  from you all to move forward with the merge requests. Thanks everyone!!

**Comment by ParaskP7** [comment]
👋 @zwarm !

> @ParaskP7 were you able to recreate this?

Unfortunately, I wasn't able to recreate this. 🤔 

> There is an issue with DomainRegistrationSource and the way it is fetching the data. The source uses continuation, but you can only resume a continuation one time and we are trying to resume continuation more than once. We end up with (1) a continuation cancel exception and (2) a illegalStateException (I believe this is the crash that @startuptester experienced) This type of logic worked perfectly fine for non-PTR situations, but once we added the refreshData support, the source falls on its face when switching sites.

This is such nice find, nice investigation Annmarie! 🌟 

> In addition to the above issue, there is also some weirdness with distinctUntilChanged(). I'm still not 100% convinced we need to remove this and I'd like to address the issues with the DomainRegistrationSource first.

Yes, I have the same feeling. I think that when everything is merged on `develop`, that is the current pull-to-refresh solution, along with the persistence and networking solution, then we can make a more educative decision on how to proceed with `distinctUntilChanged()`. 💯 

> TLDR; Since the next PR #15620 builds off of this PR, but also changes how the sources are managed. I'd like to suggest that we merge this PR and the next.

I agree. 👍 

> Following that I can open a brand new issue to address the DomainRegistrationSource changes. It's becoming too unwieldy to keep this PR up-to-date with develop + everything PTR related is behind the feature flag.

Yes please. 🥇 

> So .. I need the 👍 from you all to move forward with the merge requests.

From my side it's a 👍 especially given the fact that this does not affect the product when the `Phase 2` feature flag is off. Also, we have a whole month to test whatever enters `develop` from now on.

> Thanks everyone!!

Thank YOU!
